docs/configuration-files.md

```diff
@@ -272,91 +276,90 @@ rules:
 
 This rule looks for usage of Django's [`FloatField`](https://docs.djangoproject.com/en/3.0/ref/models/fields/#django.db.models.FloatField) model when storing currency information. `FloatField` can lead to rounding errors and should be avoided in favor of [`DecimalField`](https://docs.djangoproject.com/en/3.0/ref/models/fields/#django.db.models.DecimalField) when dealing with currency. Here the `pattern-where-python` operator allows us to utilize the Python `in` statement to filter findings that look like currency.
 
-## Metavariable matching
+## Metavariable Matching
 
-### Metavariables in logical inclusions
+Metavariable matching operates differently for logical AND (`patterns`)
+and logical OR (`pattern-either`) parent operators. Note that the behavior is
+consistent across all child operators: `pattern`, `pattern-not`,
+`pattern-regex`, `pattern-inside`, `pattern-not-inside`.
 
-Patterns' matched metavariable values are enforced to be identical when performing logical inclusion operations (`patterns`, `pattern-inside`) on matches.
+### Metavariables in Logical ANDs
```

Comment:
The reason I used "inclusion" here is that it covers both `patterns` and `pattern-inside`.

Not sure what the best "boolean" analogy is here.

Reply:
Maybe I'm misunderstanding how this works. My understanding is that metavariable matching occurs across `patterns` and `pattern-either` operators. It's operates differently in these cases, but the behavior is consistent across all child operators: `pattern`, `pattern-not`, `pattern-regex`, `pattern-inside`, `pattern-not-inside`.

Is `pattern-inside` doing something special compared to `pattern`, `pattern-not`, etc?

Reply:
Hmm, `pattern-inside` is not a simple boolean AND when used inside `patterns` (IIRC, it is not commutative, for instance). This is why I've been distinguishing it from other child node of `patterns`. 🤷

Reply:
In my head, this is a logical AND:
```yaml
patterns:
  - pattern: ...
  - pattern: ...
```

and this is not:
```
patterns:
  - pattern-inside: ...
  - pattern: ...
```

Reply:
Ahh, I see what you're saying. When using `pattern-inside`, `A & B` is not necessarily the same as `B & A`, so it's not truly a logical AND operation. Regardless, I still think the best way to word this to an average/entry-level engineer is as an AND operation for two reasons:

* I think user's will understand this better as an AND operation. For example, in Python these two statements are not equivalent in the same sense: `False and operation_with_side_effect()` vs. `operation_with_side_effect() and False`. Yet Python still describes this as an `and`.
* We describe this as an AND operation elsewhere in `semgrep`, and that's fundamentally how we think about the `patterns` operator. `pattern & pattern-inside` vs. `pattern-inside & pattern` are not the same, but I believe that's how people intuitively think about it when using `patterns`.

Maybe this is indicative of a broader docs need around the `inside` operators. I.e. we should state these non-equivalences upfront vs. doing it piecemeal in the metavariable matching section. Although I haven't heard any confusion from user's around this yet, so maybe it's not as big of an issue as we think it is. I think user's will intuitively understand that two code snippets cannot simultaneously be inside each other.

More broadly, I still think the best way to split these docs up is: under `patterns` the metavariables must match, and under `pattern-either` they don't have to match. After you understand that, then things should be consistent with other `semgrep` behavior.

Reply:
Also, I might just be wrong... at least in the simple case:

- https://semgrep.live/PekN?version=develop
- https://semgrep.live/JDdo?version=develop

🤷 Given this I think my mental model matches up with yours a lot better.

Reply:
Oh interesting, I wouldn't've assumed that would work either. Well, the other upside of relentlessly documenting things is we get to learn how things work too :)